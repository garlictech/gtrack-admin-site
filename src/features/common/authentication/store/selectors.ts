import { Injectable } from '@angular/core';

import _get from 'lodash-es/get';

import { createFeatureSelector, createSelector, MemoizedSelector } from '@ngrx/store';

import { DeepstreamSelectors } from '@bit/garlictech.angular-features.common.deepstream-ngx/src/store/selectors';

import { User } from '../interfaces';
import { Auth, AuthenticationState, featureName, JwtAuthState } from './state';

const selectAuthFeature = createFeatureSelector<AuthenticationState>(featureName);

const selectJwtFeature = createSelector(
  selectAuthFeature,
  (state: AuthenticationState) => state.jwtAuth
);

const selectAuth = createSelector(
  selectJwtFeature,
  (state: JwtAuthState) => state.auth
);

export const selectUser = createSelector(
  selectAuth,
  (state: Auth) => _get(state, 'user')
);

export const selectRole = createSelector(
  selectUser,
  (user: User) => _get(user, 'roles[0]')
);
export const selectUserId = createSelector(
  selectUser,
  (user: User) => _get(user, 'id')
);
export const loggedOut = createSelector(
  selectUser,
  user => !user
);

@Injectable({
  providedIn: 'root'
})
export class AuthenticationSelectors {
  auth: MemoizedSelector<object, Auth>;
  user: MemoizedSelector<object, User>;
  loggingIn: MemoizedSelector<object, boolean>;
  jwtLoggingIn: MemoizedSelector<object, boolean>;
  loggedIn: MemoizedSelector<object, boolean>;
  loggedOut: MemoizedSelector<object, boolean>;
  magicLinkEmailSent: MemoizedSelector<object, boolean>;
  loginFailed: MemoizedSelector<object, boolean>;
  role;
  selectedRole;
  token: MemoizedSelector<object, string>;
  termsAccepted: MemoizedSelector<object, boolean>;
  loginRefused: MemoizedSelector<object, boolean>;
  selectUserId: MemoizedSelector<object, string>;

  constructor(private readonly _deepstreamSelectors: DeepstreamSelectors) {
    const selectUiFeature = createSelector(
      selectAuthFeature,
      (state: AuthenticationState) => state.uiState
    );

    this.auth = selectAuth;
    this.user = selectUser;

    this.jwtLoggingIn = createSelector(
      selectJwtFeature,
      jwtState => jwtState.loggingIn
    );

    this.loggingIn = createSelector(
      this.jwtLoggingIn,
      this._deepstreamSelectors.loggingIn,
      (jwtLoggingIn, dsState) => jwtLoggingIn || dsState
    );

    this.loggedOut = loggedOut;

    this.token = createSelector(
      selectAuth,
      state => _get(state, 'token')
    );

    this.loggedIn = createSelector(
      this.token,
      this._deepstreamSelectors.loggedIn,
      (token, dsState) => !!token && dsState
    );

    this.magicLinkEmailSent = createSelector(
      selectJwtFeature,
      state => state.emailSent
    );

    this.loginFailed = createSelector(
      selectJwtFeature,
      this._deepstreamSelectors.failed,
      (jwtState, deepstreamFailed) => jwtState.failed || !!deepstreamFailed
    );

    this.role = selectRole;
    this.selectUserId = selectUserId;

    this.selectedRole = createSelector(
      selectUiFeature,
      state => state.selectedRole
    );

    this.termsAccepted = createSelector(
      selectUiFeature,
      state => state.termsAccepted
    );

    this.loginRefused = createSelector(
      selectUiFeature,
      state => state.loginRefused
    );
  }
}
