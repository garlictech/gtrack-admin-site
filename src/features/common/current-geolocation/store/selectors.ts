import { createFeatureSelector, createSelector } from '@ngrx/store';
import { CurrentGeolocationState, featureName } from './state';

export const selectFeature = createFeatureSelector<CurrentGeolocationState>(featureName);

export const selectTracking = createSelector(
  selectFeature,
  (state: CurrentGeolocationState) => state.tracking
);

export const selectCurrentLocation = createSelector(
  selectFeature,
  (state: CurrentGeolocationState) => state.currentLocation
);
