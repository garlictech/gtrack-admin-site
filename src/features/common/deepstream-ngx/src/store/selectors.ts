import { Inject, Injectable } from '@angular/core';
import { createFeatureSelector, createSelector, MemoizedSelector, select, Store } from '@ngrx/store';
import { filter, map } from 'rxjs/operators';
import { EXTERNAL_DEEPSTREAM_DEPENDENCIES, ExternalDeepstreamDependencies } from '../lib/externals';
import { DeepstreamState, EDeepstreamState, featureName } from './state';

import _get from 'lodash-es/get';

import { combineLatest } from 'rxjs';

export interface SelectorUserData {
  userId?: string;
  role: string;
}

@Injectable()
export class DeepstreamSelectors {
  loggingIn: MemoizedSelector<object, boolean>;
  loggedIn: MemoizedSelector<object, boolean>;
  loggedOut;
  failed;
  getPermissionRecord;
  userData: MemoizedSelector<any, SelectorUserData>;
  permissions;
  permissionRecordName;
  isOpenUser;
  private readonly _externals: ExternalDeepstreamDependencies;

  constructor(@Inject(EXTERNAL_DEEPSTREAM_DEPENDENCIES) externals, private readonly _store: Store<any>) {
    this._externals = externals;

    const selectFeature = createFeatureSelector<DeepstreamState>(featureName);

    this.permissions = createSelector(
      selectFeature,
      this._externals.selectors.getUserRole,
      (state, role) => _get(state, 'permissionRecord')
    );

    this.loggingIn = createSelector(
      selectFeature,
      this.permissions,
      (state, permissions) =>
        state.state === EDeepstreamState.LOGGING_IN || (_get(state, 'auth.id') !== 'open' && permissions === null)
    );

    this.permissionRecordName = createSelector(
      selectFeature,
      state => _get(state, 'auth.permissionRecord')
    );

    this.loggedIn = createSelector(
      selectFeature,
      this.permissions,
      (state, permissions) => state.state === EDeepstreamState.LOGGED_IN && !!permissions
    );

    this.loggedOut = createSelector(
      selectFeature,
      state => state.state === EDeepstreamState.LOGGED_OUT
    );

    this.failed = createSelector(
      selectFeature,
      state => state.failure
    );

    this.userData = createSelector(
      this._externals.selectors.getUserId,
      this._externals.selectors.getUserRole,
      (userId, role) => ({ userId, role })
    );

    this.isOpenUser = createSelector(
      selectFeature,
      state => _get(state, 'auth.id') === 'open'
    );

    this.getPermissionRecord = combineLatest(
      this._store.pipe(select(this._externals.selectors.getUserRole)),
      this._store.pipe(
        select(this.permissions),
        filter(permissions => !!permissions)
      )
    ).pipe(map((role, permissions) => _get(permissions, role) || {}));
  }
}
