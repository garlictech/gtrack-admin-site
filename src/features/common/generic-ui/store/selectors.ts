import { createFeatureSelector, createSelector } from '@ngrx/store';

import { featureName, State } from './state';

export const selectFeature = createFeatureSelector<State>(featureName);

export const selectProgressSpinnerOn = createSelector(
  selectFeature,
  (state: State) => state.progressSpinnerOn
);
export const selectProgressSpinnerText = createSelector(
  selectFeature,
  (state: State) => state.progressSpinnerText
);
