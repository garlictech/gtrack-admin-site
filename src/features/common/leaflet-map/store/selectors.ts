import { createFeatureSelector, createSelector } from '@ngrx/store';
import { featureName, State } from './state';

export const featureSelector = createFeatureSelector<State>(featureName);

export const getMapId = createSelector(
  featureSelector,
  (state: State) => state.mapId
);

export const selectFeatureId = createSelector(
  featureSelector,
  (state: State) => state.featureId
);
