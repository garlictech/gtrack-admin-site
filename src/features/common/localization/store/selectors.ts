import { createFeatureSelector, createSelector } from '@ngrx/store';

import _get from 'lodash-es/get';

import { domain, LocalizationState } from './state';

const selectFeature = createFeatureSelector<LocalizationState>(domain);

export const currentLanguage = createSelector(
  selectFeature,
  state => _get(state, 'actualLanguage', undefined)
);
