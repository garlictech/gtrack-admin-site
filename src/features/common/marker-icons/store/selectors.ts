import { createFeatureSelector, createSelector } from '@ngrx/store';

import { EIconStyle } from '../enums';
import { getSvgContent } from '../pure';
import { svgCircleContentAdapter, svgIconContentAdapter, svgMarkerContentAdapter } from './reducer';
import { featureName, State, SvgContent } from './state';

export const featureSelector = createFeatureSelector<State>(featureName);

const svgIconSelector = createSelector(
  featureSelector,
  (state: State) => state.icons
);
export const getAllSvgIcons = svgIconContentAdapter.getSelectors(svgIconSelector).selectAll;
export const getAllSvgIconsCount = svgIconContentAdapter.getSelectors(svgIconSelector).selectTotal;

export const getIcon = (type: string, encoded: boolean, iconStyle: EIconStyle) =>
  createSelector(
    getAllSvgIcons,
    (svgContents: Array<SvgContent>) => getSvgContent(type, svgContents, encoded, iconStyle)
  );

const svgMarkerSelector = createSelector(
  featureSelector,
  (state: State) => state.markers
);
export const getAllSvgMarkers = svgMarkerContentAdapter.getSelectors(svgMarkerSelector).selectAll;
export const getAllSvgMarkersCount = svgMarkerContentAdapter.getSelectors(svgMarkerSelector).selectTotal;

export const getMarker = (type: string, encoded: boolean, iconStyle: EIconStyle) =>
  createSelector(
    getAllSvgMarkers,
    (svgContents: Array<SvgContent>) => getSvgContent(type, svgContents, encoded, iconStyle)
  );

const svgCircleSelector = createSelector(
  featureSelector,
  (state: State) => state.circles
);
export const getAllSvgCircles = svgCircleContentAdapter.getSelectors(svgCircleSelector).selectAll;
export const getAllSvgCirclesCount = svgCircleContentAdapter.getSelectors(svgCircleSelector).selectTotal;

export const getCircle = (type: string, encoded: boolean, iconStyle: EIconStyle) =>
  createSelector(
    getAllSvgCircles,
    (svgContents: Array<SvgContent>) => getSvgContent(type, svgContents, encoded, iconStyle)
  );
