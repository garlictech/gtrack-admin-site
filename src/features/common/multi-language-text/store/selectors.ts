import { Injectable } from '@angular/core';
import { createFeatureSelector, createSelector, select, Store } from '@ngrx/store';
import _get from 'lodash-es/get';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { LocalizationState } from '@bit/garlictech.angular-features.common.localization';

const selectFeature = createFeatureSelector<LocalizationState>('language');

export const currentLanguage = createSelector(
  selectFeature,
  state => _get(state, 'actualLanguage', undefined)
);

export const descriptionLanguageList = createSelector(
  selectFeature,
  state => _get(state, 'descriptionLanguageList', [])
);

// The various property selectors. Consider using the reselect package.
@Injectable()
export class LocalizeSelectors {
  constructor(private readonly _store: Store<any>) {}

  getCurrentLanguage(): Observable<string> {
    return this._store.pipe(select(currentLanguage));
  }

  getDescriptionLanguageList(): Observable<Array<string>> {
    return this._store.pipe(select(descriptionLanguageList));
  }

  getLanguageSettings(): Observable<LocalizationState> {
    return this._store.pipe(
      select(selectFeature),
      map(state => ({
        actualLanguage: _get(state, 'language', 'en_US'),
        descriptionLanguageList: _get(state, 'descriptionLanguageList', [])
      }))
    );
  }
}
