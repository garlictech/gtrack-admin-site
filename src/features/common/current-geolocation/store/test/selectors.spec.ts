import * as fromSelectors from '../selectors';
import { createState } from './utils';

const testCases = [
  {
    name: 'selectCurrentLocation',
    selector: fromSelectors.selectCurrentLocation,
    state: createState()
  },
  {
    name: 'selectTracking',
    selector: fromSelectors.selectTracking,
    state: createState()
  }
];

testCases.forEach(({ name, selector, state }) => {
  test(`${name} with input ${JSON.stringify(state)}`, () => {
    expect(selector(state)).toMatchSnapshot();
  });
});
