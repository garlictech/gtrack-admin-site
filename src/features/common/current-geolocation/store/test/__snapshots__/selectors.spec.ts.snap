// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectCurrentLocation with input {"features.common.current-geolocation":{"tracking":false}} 1`] = `undefined`;

exports[`selectTracking with input {"features.common.current-geolocation":{"tracking":false}} 1`] = `false`;
